### PR TITLE
Support worker termination by collecting sum stats on exit.

### PR DESCRIPTION
**Pull request by jdmarshall**
This is the last PR required for the next semver-major release of the Prometheus Client. It addresses #803, via code changes, exposing additional lifecycle events for managing worker lifetimes, and documentation changes to warn of corner cases.

This change creates a retained mode for defunct workers where all of their final telemetry for 'sum' metrics is recorded in the aggregator process and aggregated down with each subsequent terminated worker. It also uses a call and response 
system so that the worker can ensure that the flushed data was acknowledged by the collector before exiting.

As part of this work, most of the 'thenables' have been removed from the `clusterMetrics()` code path, in favor of async/await which should also reduce the surface area of incompatibility with third party Promise libraries. And in general the code complexity of `clusterMetrics` has been reduced close to the essential complexity of this code, to satisfy Kernighan's Law with respect to spotting bugs in the implementation. 

Pains are taken to make sure that the `workers` and `historicalMetrics` fields are only ever read or written on the same tick, so that any shutdown operations that overlap with scraping don't smash into each other and cause data artifacts. You either get a scrape because the workers all replied before shutting down, or you get a transient timeout error that corrects itself on the subsequent collection.

The collector thread only collects 'sum' metric types - a dead process cannot participate in average, max, or min calculations any longer and those outliers shouldn't be retained once the process exits. However counts and histograms must be kept in aggregate to prevent sawtooth patterns in the collected data due to elision of previously-reported data.

This PR also reworks a number of tests that relied on `delay()` to instead watch the appropriate EventEmitters to watch for events that are being emitted asynchronously. There are a couple of cases where tests have around a 5-10% failure rate due to timing issues and if you run the suite over and over again, you will see false positives intermittently. Most of these should now be corrected. 

Because this is trying to keep two implementations in lockstep, it is unfortunately again a rather large PR, but hopefully not difficult to comprehend.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -45,16 +45,16 @@ This release marks our first release under the Prometheus umbrella.
 - perf: Use faster `process.memoryUsage.rss()` API for resident memory collection (30-40% more ops/sec)
 - fix: Browser compatibility for Gauge.startTimer()
 - ci: Run benchmarks for pull requests
-- ci: switch out deprecated benchmark-regression library for replacement
```
Comment (krajorama) [inline CHANGELOG.md:48]: should this be removed?

File: CHANGELOG.md [hunk 2]
```diff
@@ -45,16 +45,16 @@ This release marks our first release under the Prometheus umbrella.
...
 - chore: Add copyright license headers and test
-- Make cluster and worker-thread metric aggregation order deterministic
+- Support for worker threads added
+- Cluster and Worker improvements:
+  - Make cluster and worker-thread metric aggregation order deterministic
+  - Graceful shutdown support, and data retention past worker termination
 - Export `MetricObject`, `MetricObjectWithValues`, `MetricValue` and `MetricValueWithName` from the TypeScript definitions
 - chore: Old label processing code marked as deprecated
-- Improve cluster support to allow workers to opt out
-- Abort cluster metric responses during process termination
```
Comment (krajorama) [inline CHANGELOG.md:57]: should these be removed?

File: Workers.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
+# Notes on collection and short-lived processes
+
+## Introduction
+
+Worker threads present a large surface area for potential loss of telemetry data. In order to avoid
+strange artifacts in your statistics, it's recommendeng that you hook the process lifecycle events
```
Comment (krajorama) [inline Workers.md]:
```suggestion
strange artifacts in your statistics, it's recommended that you hook the process lifecycle events
```

File: lib/cluster.js [hunk 4]
```diff
@@ -60,76 +65,136 @@ class AggregatorRegistry extends Registry {
...
 			.filter(worker => worker.isConnected())
 			.sort((left, right) => left.id - right.id);
 
-		return new Promise((resolve, reject) => {
-			let settled = false;
-			function done(err, result) {
-				if (settled) return;
-				settled = true;
+		if (orderedWorkers.length === 0) {
+			debug('No workers found for requestId', requestId);
```
Comment (krajorama) [inline lib/cluster.js:75]: I wonder if this is necessary. Could we have  a situation when the application is idling and this keeps getting called?

File: test/clusterTest.js [hunk 5]
```diff
@@ -150,6 +171,168 @@ describe.each([
...
+	});
+
+	describe('shutdown()', () => {
+		let AggregatorRegistry;
+		let listener;
+		let discovery;
+
+		beforeEach(() => {
+			jest.resetModules();
+			AggregatorRegistry = require('../lib/worker');
```
Comment (krajorama) [inline test/clusterTest.js]: Seems like copy paste bug?

```suggestion
			AggregatorRegistry = require('../lib/cluster');
```

File: test/registerTest.js [hunk 6]
```diff
@@ -340,7 +340,36 @@ describe('Register', () => {
...
+
+				expect(output.length).toEqual(1);
+				expect(output[0].name).toEqual('test_metric');
+				expect(output[0].get).toBeInstanceOf(Function);
+			});
+
+			describe('with aggregator argument', () => {
+				it('should filter out other aggregators', async () => {
+					const max = getMetric('max_metric');
+					max.aggregater = 'max';
```
Comment (krajorama) [inline test/registerTest.js]: typo:

```suggestion
					max.aggregator = 'max';
```

File: Workers.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+['SIGHUP', 'SIGINT', 'SIGQUIT', 'SIGTERM'].forEach(sig => {
+  process.on(sig, async () => {
+    await registry.shutdown();
+    process.exit(0);
+  });
+});
+```
+
+See [the example](examples/workerTest.js) for a complete rundown, including the parent process
```
Comment (krajorama) [inline Workers.md:70]: typo

```suggestion
See [the example](examples/worker.js) for a complete rundown, including the parent process
```

File: README.md [hunk 8]
```diff
@@ -41,13 +44,23 @@ for an example.)
...
+cardinality situation, which the Aggregator is generally meant to avoid.
 
 Metrics are aggregated from the global registry by default. To use a different
 registry, call
 `client.AggregatorRegistry.setRegistries(registryOrArrayOfRegistries)` from the
 worker processes.
 
+#### Process Lifecycle
+
+The `shutdown()' method is provided to help cleanly shut down the application while
```
Comment (krajorama) [inline README.md]: typo:

```suggestion
The `shutdown()` method is provided to help cleanly shut down the application while
```

File: lib/cluster.js [hunk 9]
```diff
@@ -293,26 +388,50 @@ function announce() {
 }
 
 /**
- * Replace any listeners with new ones.
+ * Look for and complain about duplicate listeners.
+ *
+ * In test and delevopment mode, this function will remove duplicate listeners.
```
Comment (krajorama) [inline lib/cluster.js]: typo:

```suggestion
 * In test and development mode, this function will remove duplicate listeners.
```

File: README.md [hunk 10]
```diff
@@ -636,3 +649,33 @@ To avoid native dependencies in this module, GC statistics for bytes reclaimed
...
+features were intended almost exclusively for compatibility with test harnesses in order to
+support unit and integration tests. It has not been hardened to support workloads where a live
+application forcibly reloads modules to for instance pick up changes from the filesystem without
+restarting the server.
+
+Using this in production will result in data loss. Lost telemetry can trigger anything from alert to
+incorrect triage of production issues when 'sum' metrics show glitches or sawtooth patterns in their
+charts.
+
+While reloadng your own code will often not also reload your dependencies, some extra care may be
```
Comment (krajorama) [inline README.md]: typo:


```suggestion
While reloading your own code will often not also reload your dependencies, some extra care may be
```

File: README.md [hunk 11]
```diff
@@ -636,3 +649,33 @@ To avoid native dependencies in this module, GC statistics for bytes reclaimed
 in each GC sweep are kept in a separate module:
 https://github.com/SimenB/node-prometheus-gc-stats. (Note that that metric may
 no longer be accurate now that v8 uses parallel garbage collection.)
+
+## Notes
+
+### Hot Reloading
+
+While it is unusual for NodeJS applications to reload modules at runtime, some more esoteric
+codebases may still due so. You should be aware that there is quite a bit of undefined behavior
```
Comment (krajorama) [inline README.md]: typo:

```suggestion
codebases may still do so. You should be aware that there is quite a bit of undefined behavior
```

File: lib/cluster.js [hunk 12]
```diff
@@ -276,6 +352,25 @@ async function primaryListener(worker, event) {
...
+
+			worker.send({
+				type: ACK,
+				requestId: event.requestId,
+			});
+
+			try {
+				const metrics = [historicMetrics, ...event.metrics];
+
+				historicMetrics = await Registry.aggregate(metrics).getMetricsAsArray();
```
Comment (krajorama) [inline lib/cluster.js]: Seems to me these are just function calls, not promises, so no `await` needed?

```suggestion
				historicMetrics = Registry.aggregate(metrics).getMetricsAsArray();
```

**Review by krajorama — COMMENTED** [review]
First pass.

The general approach makes sense to me, but will need another pass to go into the details.

File: lib/cluster.js (hunk 4 above)
Comment (jdmarshall) [reply lib/cluster.js]: In debug mode you’re not idling you’re debugging.

File: CHANGELOG.md (hunk 1 above)
Comment (jdmarshall) [reply CHANGELOG.md:48]: There’s two years of updates in chronological order in one release we don’t need to announce we added a library and upgraded it twice. 

> chore: replace benchmark-regression dependency with faceoff 

Is still in there

File: CHANGELOG.md (hunk 2 above)
Comment (jdmarshall) [reply CHANGELOG.md:57]:
```
 - Cluster and Worker improvements:
  - Make cluster and worker-thread metric aggregation order deterministic
  - Graceful shutdown support, and data retention past worker termination
```

I suppose “Improve cluster support to allow workers to opt out” got dropped

File: lib/cluster.js (hunk 4 above)
Comment (jdmarshall) [reply lib/cluster.js:75]: Not while you’re debugging. Knowing that there are no workers is information. Could be a bug in announcements I missed. 

Mostly this helped tremendously getting the mocks right in the tests. They were a mess, particularly in a couple of the contributed tests. I literally spent the majority of my time on this fixing tests I broke with the added events.

File: lib/cluster.js (hunk 12 above)
Comment (jdmarshall) [reply lib/cluster.js]: Hmm. That’s odd. One is sync and the rest are async. However did we manage that?

But I should double check that collect() is being called in the worker because I bet I missed one. 

And this needs to be fixed in both.

Also there's an unhandledRejectionError surface area here that I missed.  This code evolved during testing and I think I ran out of steam before it was finished.

File: test/clusterTest.js (hunk 5 above)
Comment (jdmarshall) [reply test/clusterTest.js]: Looks like I missed one, yep.  Trying to make sure they test the same corner cases in the same way, so if someone adds a new feature...

**Review by krajorama — COMMENTED** [review]
lgtm, llm claims some flakiness/cross pollination in test only - I'll open a PR for that afterwards, I need to understand it myself.

Please resolve the remaining small typo/fixe and I can approve.

**Review by krajorama — APPROVED** [review]
LGTM, I've also done a simple smoke test locally. should be good for RC
